pkg/ruler/remote/querier.go

```diff
@@ -0,0 +1,195 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package remote
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/textproto"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/weaveworks/common/httpgrpc"
+
+	"github.com/grafana/mimir/pkg/util/httpgrpcutil"
+	"github.com/grafana/mimir/pkg/util/spanlogger"
+	"github.com/grafana/mimir/pkg/util/version"
+)
+
+const (
+	readEndpointPath  = "/api/v1/read"
+	queryEndpointPath = "/api/v1/query"
+
+	mimeTypeFormPost = "application/x-www-form-urlencoded"
+
+	statusError = "error"
+)
+
+var userAgent = fmt.Sprintf("mimir/%s", version.Version)
+
+// Middleware provides a mechanism to inspect outgoing Querier requests.
+type Middleware func(ctx context.Context, req *httpgrpc.HTTPRequest) error
+
+// Querier executes read operations against a httpgrpcutil.RoundTripper.
+type Querier struct {
+	transport      httpgrpcutil.RoundTripper
+	middlewares    []Middleware
+	promHTTPPrefix string
+	logger         log.Logger
+}
+
+// New creates and initializes a new Querier instance.
+func New(
+	transport httpgrpcutil.RoundTripper,
```

Comment:
Why don't we use `httpgrpc.HTTPClient` directly (which is already a interface and can be easily mocked in the tests), but introduce new interface in the middle?

Reply:
Also, in this case we could also get rid of `httgrpcutil.RoundTripper` type. wdyt?

Reply:
I agree.